modules/hal_nordic/nrfs/backends/nrfs_backend_ipc_service.c

```diff
@@ -0,0 +1,341 @@
+
+static K_SEM_DEFINE(ipc_init_done_sem, 0, 1);
+static K_SEM_DEFINE(ipc_sysctrl_bound_sem, 0, 1);
+static K_EVENT_DEFINE(ipc_connected_event);
+
+#define IPC_INIT_DONE_EVENT (0x01)
+
+struct ipc_channel_config_t {
+	const struct device *ipc_instance;
+	struct ipc_ept_cfg *endpoint_config;
+	struct k_sem *bound_sem;
+	struct ipc_ept ipc_ept;
+	atomic_t status;
+	bool enabled;
+};
+
+/**
+ * @brief nrfs backend error handler
+ *
+ * @param error_id The id of an error to handle.
+ * @param error additional error code if needed, if not needed use 0.
+ * @param fatal true if fatal error and needs special handling
+ */
+__weak void nrfs_backend_error_handler(enum nrfs_backend_error error_id, int error, bool fatal)
+{
+	switch (error_id) {
+	case NRFS_ERROR_EPT_RECEIVE_QUEUE_ERROR:
+		LOG_ERR("Receive queue error: %d", error);
+		break;
+
+	case NRFS_ERROR_EPT_RECEIVE_DATA_TOO_LONG:
+		LOG_ERR("Received data is too long. Config error.");
+		break;
+
+	case NRFS_ERROR_IPC_CHANNEL_INIT:
+		LOG_ERR("IPC channel init failed with error %d", error);
+		break;
+
+	case NRFS_ERROR_SEND_DATA:
+		LOG_ERR("Send message to sysctrl failed with error: %d", error);
+		break;
+
+	case NRFS_ERROR_NO_DATA_RECEIVED:
+		LOG_ERR("No data in received message!");
+		break;
+
+	case NRFS_ERROR_IPC_OPEN_INSTANCE:
+		LOG_ERR("IPC open instance failure with error: %d", error);
+		break;
+
+	case NRFS_ERROR_IPC_REGISTER_ENDPOINT:
+		LOG_ERR("IPC register endpoint failure with error: %d", error);
+		break;
+
+	case NRFS_ERROR_BACKEND_NOT_CONNECTED:
+		LOG_ERR("Backend not connected to sysctrl.");
+		break;
+
+	default:
+		LOG_ERR("Undefined error id: %d, error cause: %d", error_id, error);
+		break;
+	}
+
+	if (fatal) {
+		nrfs_backend_fatal_error_handler(error_id);
+	}
+}
+
+static void ipc_sysctrl_ept_bound(void *priv)
+{
+	LOG_DBG("Bound to sysctrl.");
+	k_sem_give(&ipc_sysctrl_bound_sem);
+}
+
+static void ipc_sysctrl_ept_recv(const void *data, size_t size, void *priv)
+{
+	__ASSERT(size <= MAX_PACKET_DATA_SIZE, "Received data is too long. Config error.");
+	if (size <= MAX_PACKET_DATA_SIZE) {
+		rx_data.channel_id = IPC_CPUSYS_CHANNEL_ID;
+		rx_data.size	   = size;
+		if (data) {
+			memcpy(rx_data.data, (uint8_t *)data, size);
+			int status = k_msgq_put(&ipc_receive_msgq, &rx_data,
+						K_USEC(CONFIG_NRFS_BACKEND_RX_QUEUE_MAX_WAIT_US));
+
+			if (status != 0) {
+				nrfs_backend_error_handler(NRFS_ERROR_EPT_RECEIVE_QUEUE_ERROR,
+							   status, true);
+			}
+
+		} else {
+			nrfs_backend_error_handler(NRFS_ERROR_NO_DATA_RECEIVED, 0, false);
+		}
+	} else {
+		nrfs_backend_error_handler(NRFS_ERROR_EPT_RECEIVE_DATA_TOO_LONG, 0, true);
+	}
+}
+
+static struct ipc_ept_cfg ipc_sysctrl_ept_cfg = {
+	.name = "ipc_to_sysctrl",
+	.cb = {
+		.bound    = ipc_sysctrl_ept_bound,
+		.received = ipc_sysctrl_ept_recv,
+	},
+};
+
+static struct ipc_channel_config_t ipc_cpusys_channel_config = {
+	.ipc_instance	 = DEVICE_DT_GET(DT_NODELABEL(ipc_to_cpusys)),
```

Comment:
That finally blocked me, I could not fix it myself quickly. Neither `ipc_to_cpusys` nor `cpuapp_cpusys_ipc` seems to exist with all the other things that belong to it. It is like a rocket without a propulsion system. Same for the lite backend.

Reply:
Yes, not all required dts files were moved to the upstream repo. I'm checking how to fix it

Reply:
Implemented here: https://github.com/zephyrproject-rtos/zephyr/pull/71265